[UI] Working draft of class statistics page

dvolpin29 opened the pull request:
**Description**

- Working draft of a new class statistics page
- new-class-stats.html replaces the old class-stats.html
- added clickable student tiles (when a student tile is clicked, an expanded pane opens to show more detail)
- at the moment the page only uses student name information

In this new html file, the teacher can now view his students in a tile format, and select a student tile to receive more information.

This updated page should help teachers see a better overview of how each student is doing in their class.

**Discussion Number: 3447**

In the discussion board number 3447, Felienne has raised the issue of how the class statistics page looks. The proposal was to simplify it for the teachers so it is more useful.

**How to test**

* no testing needed at the moment

**Questions**

Tried unsuccessfully to use new tailwind css, such as hover, active with specific colours that do not exist in the static/css/generated.full.css file. I am not sure how I can add new tailwind css to be able to use it.

Felienne commented:
Looks nice @dvolpin29!!

<img width="1119" alt="image" src="https://user-images.githubusercontent.com/1003685/230553595-77d0734c-e2b5-4ffa-a867-afc380f729e0.png">

A few notes:

* Shall we keep the old page alive for now also? Just have this next to it is probably the best until we are ready to make a switch. We can make an extra button for this new page (just like #4166 also does) so that we can have them side by side for a while.

* Have you read the docs about the dev css and does that solve the issue?

* We do like to have tests for new features to make sure they work and don't break afterwards. Can you add a few Cypress tests (in due time, can also be done when we are more sure what the feature will look like)

dvolpin29 commented:
- added an additional button in **class-overview.html** for the new class statistics page. 
-  **new-class-stats.html** page uses a new function from  **statistics.py** called _render_new_class_stats_
    -  _render_new_class_stats_ -> extracts additional student information into the student tile, and expanded student tile

Still need to see how to use more tailwind css, instead of JavaScript.

rix0rrr commented on templates/new-class-stats.html:
```diff
@@ -0,0 +1,87 @@
+
+</div>
+
+<script>
```
As you said yourself already, some of this JavaScript's effects can probably be replaced with CSS. Please avoid using  JavaScript unless absolutely necessary. If you do add JavaScript, please don't add it in inline. JavaScript should go into the `static/js/*.ts` files (pick a file that makes sense).

dvolpin29 replied:
Yes, I added the javaScript file! I will slowly remove some javaScript functions, once I have found matching tailwind CSS functionality that can replace it.

rix0rrr reviewed (commented):
> Tried unsuccessfully to use new tailwind css, such as hover, active with specific colours that do not exist in the static/css/generated.full.css file. I am not sure how I can add new tailwind css to be able to use it.

After you add something like `class="hover:bg-green-300"`, run the following command in your terminal:

```
./build-tools/heroku/tailwind/generate-development-css
```

(Described [here](https://github.com/hedyorg/hedy/blob/main/CONTRIBUTING.md#working-on-the-web-front-end-in-tailwind))

rix0rrr commented on static/js/student-tile.js:
Please make this a TypeScript source file.

rix0rrr commented:
Ha, I spoke too soon. Give me a moment :)

rix0rrr commented:
Okay I missed one part in my previous commit -- I hadn't noticed the "selected" styles were also being done with JavaScript, by putting inline styles on the `<p>` elements. This messed up the CSS-applied styles a little, because inline styles win out over styles applied using CSS.

In this last commit I did the following:

Introduced a new rule in `styles.css`:

```css
.student-tile.selected {
  @apply bg-green-400 text-white;
}
```

This says that any element that has BOTH the class `student-tile` *and* `selected` gets the following styles applied (green background and white text).

After that, the JavaScript doesn't have to set and remove the actual colors anymore. All it needs to do is add and remove the `selected` class from the `student-tile`s:

```js
    $('.student-tile').removeClass('selected');
    $(event.target).closest('.student-tile').addClass('selected');
```

(I made it slightly more complicated to make it so you can also *deselect* a student tile -- I felt that's a nice affordance that makes it possible to restore the page to the state the user finds it in).

rix0rrr commented:
(P.S.: Not that this is necessarily of great help to you now, but all this CSS-manipulating JavaScript you're writing right now could be replaced with a bit of hyperscript once this PR gets merged: https://github.com/hedyorg/hedy/pull/4224. Keep an eye out for it!)

dvolpin29 commented:
> (P.S.: Not that this is necessarily of great help to you now, but all this CSS-manipulating JavaScript you're writing right now could be replaced with a bit of hyperscript once this PR gets merged: #4224. Keep an eye out for it!)

Thank you for all the help, I will keep an eye on it!

dvolpin29 commented:
> Okay I missed one part in my previous commit -- I hadn't noticed the "selected" styles were also being done with JavaScript, by putting inline styles on the `<p>` elements. This messed up the CSS-applied styles a little, because inline styles win out over styles applied using CSS.
> 
> In this last commit I did the following:
> 
> Introduced a new rule in `styles.css`:
> 
> ```css
> .student-tile.selected {
>   @apply bg-green-400 text-white;
> }
> ```
> 
> This says that any element that has BOTH the class `student-tile` _and_ `selected` gets the following styles applied (green background and white text).
> 
> After that, the JavaScript doesn't have to set and remove the actual colors anymore. All it needs to do is add and remove the `selected` class from the `student-tile`s:
> 
> ```js
>     $('.student-tile').removeClass('selected');
>     $(event.target).closest('.student-tile').addClass('selected');
> ```
> 
> (I made it slightly more complicated to make it so you can also _deselect_ a student tile -- I felt that's a nice affordance that makes it possible to restore the page to the state the user finds it in).

I agree, it is definitely better now! Thank you for the updates, I will work with this instead.

dvolpin29 commented:
> To download my new changes to your branch, you can run `git pull`. I also changed the TypeScript, but I didn't commit the regenerated JavaScript `appbundle.js`, so you'll have to do that on your end to continue working on this locally (`./build-tools/heroku/generate-typescript`).

Small question, I have changed the student-tile.js to a typeScript file, and generated the type script using ./build-tools/heroku/generate-typescript. Should I, commit also appbundle.js? or keep it locally?

Felienne commented:
> Small question, I have changed the student-tile.js to a typeScript file, and generated the type script using ./build-tools/heroku/generate-typescript. Should I, commit also appbundle.js? or keep it locally?

Commit it! This makes is as easy as possible for people checking your branch/PR to run it, since they will then not have to compile. But if you don't commit it, it will still be added on merge, so forgetting is no big deal.

dvolpin29 reviewed (commented):
Not sure, why this 2nd commit happned.

dvolpin29 commented:
First draft of the new page is finalised.

dvolpin29 commented:
Working to finalise the second statistics page about each student. Here are the details of the changes that were made.

Completed:
- drop down menu to choose between version 1 and version 2 of the new statistics page for the teachers page
- version 2 includes a bar chart that shows how many programs the student has run on each level (still need to do small modifications)

Todo:
- Add a linear chart to show how each student performed on the quizzes on each level

Felienne commented on static/js/student-tile.ts:
```diff
@@ -0,0 +1,36 @@
+    const highest_level_quiz_score = student.highest_level_quiz_score;
+
+    let bigTile = document.getElementById('expanded-student-tile')!;
+    bigTile.innerHTML =
```
This should be done with either HTMX or with a flask/jinja template and not directly in the typescript.

dvolpin29 replied:
issue fixed.

jpelay replied:
Sorry Dvolpin! But I don't see where this issue was fixed? If you need help with HTMX or flask you can hit me up!

dvolpin29 replied:
I will try to fix to work on this tomorrow, if I get stuck on it, I will ping you! Thanks for the various inputs!

Felienne commented:
The new views look really nice @dvolpin29! They are fine as a prototype, but the code needs a bit more work to be able to be merged (but to use in a demo, this will be fine)

dvolpin29 commented on static/js/message-translations.ts:
```diff
@@ -91,7 +91,7 @@ export const TRANSLATIONS: Record<string, Record<MessageKey, string>> = {
     "Transpile_warning": "Vigila!",
     "Unsaved_Changes": "Tens un programa sense desar. Vols marxar sense desar-lo?",
-    "copy_link_to_share": "Copia l'enllaç per compartir",
+    "copy_link_to_share": "Copy link to share",
```
issue fixed, merged main branch, and discarded the changes I made.

jpelay commented on website/statistics.py:
```diff
@@ -362,3 +515,101 @@ def get_general_class_stats(students):
+        if len(quizzes_ran_per_level[level - 1]) == 1:
+            avg_quizzes_ran_per_level.append(quizzes_per_level[0])
+        if len(quizzes_ran_per_level[level - 1]) > 1:
+            average_quiz_score_per_level = 0
+            for index in range(1, len(quizzes_per_level) + 1):
+                average_quiz_score_per_level += quizzes_per_level[index - 1]
+
+            average_quiz_score_per_level /= (len(quizzes_ran_per_level[level - 1]))
+            avg_quizzes_ran_per_level.append(int(average_quiz_score_per_level))
+
+    return avg_quizzes_ran_per_level
```
Please review this code, because I tried running some quiz attemps and it only computes one attempt. Also, for taking the average you can do something like `sum(quizzes_per_level)/len(quizzes_per_level)`

dvolpin29 replied:
Issue fixed, it seems that Hedy database saves quizzes that are completed in the same week differently than if they are done in different weeks.

jpelay commented on website/statistics.py:
```diff
@@ -362,3 +515,101 @@ def get_general_class_stats(students):
+    num_quizzes = 0
+    total_quiz_scores = 0
+    for i in range(1, len(average_quizzes_ran_per_level) + 1):
+        if average_quizzes_ran_per_level[i - 1] != 0:
```
Why are you starting from 1 and then removing 1 in the index? Maybe there's a valid reason! It's just a little convoluted to me.

dvolpin29 replied:
Oh, I think that is by mistake. I changed it now.

dvolpin29 commented on static/js/index.ts:
```diff
@@ -22,3 +22,5 @@ export * from './teachers';
 export * from './initialize';
 export * from './debugging';
+export * from  './student-tile';
+export * from  './student-tile-chart';
```
I have removed `export * from  './student-tile-chart';` and `export * from  './student-tile';`. I added all the appropriate functions to statistics.ts.

dvolpin29 commented on website/statistics.py:
```diff
@@ -46,11 +47,163 @@ def render_class_stats(self, user, class_id):
+
+        students = []
+        for student_username in class_.get("students", []):
+            print(f'Student: {student_username}')
```
Removed!

dvolpin29 commented on website/statistics.py:
```diff
@@ -46,11 +47,163 @@ def render_class_stats(self, user, class_id):
+        if not class_ or (class_["teacher"] != user["username"] and not is_admin(user)):
+            return utils.error_page(error=404, ui_message=gettext("no_such_class"))
+
+        levels = []
+        for i in range(1, hedy.HEDY_MAX_LEVEL + 1):
+            levels.append(i)
```
Makes sense, still getting used to using more pythonic "style".

dvolpin29 commented on website/statistics.py:
```diff
@@ -362,3 +515,101 @@ def get_general_class_stats(students):
+
+# append the quizzes obtained on each level
+def find_quizzes_per_level(quizzes_per_level, quizzes, level):
+    quizzes_per_level.append([])
```
I will check into that, I am not as familiar with creating dictionaries on python.

dvolpin29 commented on website/statistics.py:
```diff
@@ -362,3 +515,101 @@ def get_general_class_stats(students):
+# append the programs obtained on each level
+def find_program_runs_per_level(program_runs_per_level, success_runs_per_level, error_runs_per_level, programs, level):
+    program_runs_per_level.append([])
+    success_runs_per_level.append([]), error_runs_per_level.append([])
```
Fixed!

Felienne reviewed (changes requested):
Some code has been changed in this PR in an incorrect way (See inline comment for one example)

The ideal course of action would be for you to:

* Start a new branch of of the current main
* Reapply your changes (either with cherry picking in git or manually)
* Make fresh PR

You can also do this in a few small steps, with a few small PRs, that might also make it easier to do.

Felienne commented:
But despite the issues, the new stats do look nice:

<img width="1161" alt="image" src="https://github.com/hedyorg/hedy/assets/1003685/bd8ac69d-08a1-4c87-80c7-c32f79b88e22">

dvolpin29 commented on app.py:
```diff
@@ -33,7 +33,7 @@
 from config import config
 from website.flask_helpers import render_template, proper_tojson, JinjaCompatibleJsonProvider
-from hedy_content import (ADVENTURE_ORDER_PER_LEVEL, KEYWORDS_ADVENTURES, ALL_KEYWORD_LANGUAGES,
+from hedy_content import (ADVENTURE_ORDER_PER_LEVEL, ALL_KEYWORD_LANGUAGES,
```
> Some code has been changed in this PR in an incorrect way (See inline comment for one example)
> 
> The ideal course of action would be for you to:
> 
> * Start a new branch of of the current main
> * Reapply your changes (either with cherry picking in git or manually)
> * Make fresh PR
> 
> You can also do this in a few small steps, with a few small PRs, that might also make it easier to do.

I have created a new PR (#4336), and tested that it works as expected. You can close this Pull Request!

jpelay commented:
Hi! I'm closing this PR since we have a new one with the exact changes that we want!
